lib/seattleflu/db/cli/command/etl/presence_absence.py

```diff
@@ -0,0 +1,385 @@
+    else:
+        LOG.warning(f"No corresponding UUID found for barcode {barcode}")
+        return barcode
+
+
+def sample_details(document: dict) -> dict:
+    """
+    Describe sample details in a simple data structure designed to be used
+    from SQL.
+    """
+    return { 
+        "sample_comment": document['sampleComment'],
+        "initial_sequence_call": document['initialProceedToSequencingCall'],
+        "proceed_to_sequence": document["sampleProceedToSequencing"]
+    }
+
+def upsert_presence_absence(db: DatabaseSession,
+                            identifier: str,
+                            sample_id: int,
+                            target_id: int,
+                            present: bool,
+                            well_details: list) -> Any:
+    """
+    Upsert presence_absence by its *identifier*.
+    """
+    LOG.debug(f"Upserting presence_absence «{identifier}»")
+
+    data = {
+        "identifier": f"NWGC_{identifier}",
+        "sample_id": sample_id,
+        "target_id": target_id,
+        "present": present,
+        "well_details": Json(well_details)
+    }
+
+    presence_absence = db.fetch_row("""
+        insert into warehouse.presence_absence (
+                identifier,
+                sample_id,
+                target_id,
+                present,
+                well_details)
+            values (
+                %(identifier)s,
+                %(sample_id)s,
+                %(target_id)s,
+                %(present)s,
+                %(well_details)s)
+
+        on conflict (identifier) do update
+            set sample_id = excluded.sample_id,
+                target_id = excluded.target_id,
+                present   = excluded.present,
+                well_details = excluded.well_details
+
+        returning presence_absence_id as id, identifier
+        """, data)
+
+        # TODO: the safest thing to do in the above sql query, on conflict, is probably not 
+        # to look at the identifier for a conflict. For example, if the lab sending us the 
+        # data wants to re-run any tests, their backend may force their identifier id to change.
+        # Instead, we may want to set a constraint that any sample_id and target_id combination
+        # should be unique, and create a conflict rule in the situation where duplicates are
+        # uploaded. 
+
+    assert presence_absence.id, "Upsert affected no rows!"
+
+    LOG.info(f"Upserted presence_absence {presence_absence.id} «{presence_absence.identifier}»")
+
+    return presence_absence
+
+
+def get_target_result(target_status: str, sample_identifier: int, target_identifier: int) -> Any:
+    """
+    Takes a given target status and its sample and target ids. Returns the decoded 
+    target_result as a boolean if the given target status is known. If the given
+    target status is an unexpected value, returns 'None' and logs an error for the 
+    given sample and target ids.
```

Comment:
This description of behavior is incorrect. The error is raised immediately and no value is returned.